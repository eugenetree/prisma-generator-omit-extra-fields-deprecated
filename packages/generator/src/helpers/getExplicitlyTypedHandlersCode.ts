import { DMMF } from "@prisma/generator-helper";
import { getDefaultInputTypeName, getModelFieldsVariableName, getPartialInputTypeName } from "./common/names";

const prismaTypesToTs: Record<string, string> = {
  'Int': 'number',
  'String': 'string',
  'Boolean': 'boolean',
  'DateTime': 'Date',
}

const getInputTypesCode = (
  { modelName, modelFields }:
      { modelName: DMMF.Model['name']; modelFields: DMMF.Model['fields'] }
): string => {
  const defaultInputTypeName = getDefaultInputTypeName(modelName);
  const partialInputTypeName = getPartialInputTypeName(modelName);

  let result: string = `type ${defaultInputTypeName} = {`;
  modelFields.forEach((field) => {
      const tsType = prismaTypesToTs[field.type];
      if (tsType !== undefined) {
          result += ` ${field.name}${field.isRequired ? '' : '?'}: ${tsType};`
      }
  })

  result += `}; \n\n type ${partialInputTypeName} = {`
  modelFields.forEach((field) => {
      const tsType = prismaTypesToTs[field.type];
      if (tsType !== undefined) {
          result += ` ${field.name}?: ${tsType};`
      }
  })

  result += '}'
  return result;
}

export const getExplicitlyTypedHandlersCode = (
  { modelName, modelFields }:
    { modelName: DMMF.Model['name']; modelFields: DMMF.Model['fields'] }
) => {
  let result = getInputTypesCode({ modelName, modelFields });
  const handlersData = [
    { functionName: `polishDefault${modelName}`, typeName: getDefaultInputTypeName(modelName) },
    { functionName: `polishPartial${modelName}`, typeName: getPartialInputTypeName(modelName) }
  ]

  for (const { functionName, typeName } of handlersData) {
    result += `\n\n export const ${functionName} = (input: ${typeName}): ${typeName} => {
        const result = {};
    
        ${getModelFieldsVariableName(modelName)}.forEach((key) => {
          result[key] = input[key];
        })
    
        return result as ${typeName};
      }`
  }
  return result;
}