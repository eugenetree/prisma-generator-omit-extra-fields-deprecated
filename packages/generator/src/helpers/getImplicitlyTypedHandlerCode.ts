import { DMMF } from "@prisma/generator-helper"

export const getImplicitlyTypedHandlerCode = (
    { modelName }: { modelName: DMMF.Model['name'] }
  ) => {
    return `export const polish${modelName} = <T>(input: T): T => {
      const result = {};
  
      ${modelName.toLowerCase()}Keys.forEach((key) => {
        result[key] = input[key];
      })
  
      return result as T;
    }`
  }