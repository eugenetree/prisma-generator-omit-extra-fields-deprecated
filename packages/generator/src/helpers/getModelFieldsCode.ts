import { DMMF } from "@prisma/generator-helper";

export const getModelFieldsCode = (
    { modelName, modelFields }:
      { modelName: DMMF.Model['name']; modelFields: DMMF.Model['fields'] }
  ) => {
    const fieldNames = modelFields.reduce((res, field) => `${res} '${field.name}',`, '')
    return `const ${modelName.toLowerCase()}Keys = [${fieldNames}]`;
  }