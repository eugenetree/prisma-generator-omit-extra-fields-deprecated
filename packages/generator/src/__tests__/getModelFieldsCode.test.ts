import { getSampleDMMF } from './__fixtures__/getSampleDMMF'
import { getModelFieldsCode } from '../helpers/getModelFieldsCode';


test('model fields generation', async () => {
	const sampleDMMF = await getSampleDMMF()

	sampleDMMF.datamodel.models.forEach((model) => {
		expect(getModelFieldsCode({modelName: model.name, modelFields: model.fields})).toMatchSnapshot(model.name);
	})
})
