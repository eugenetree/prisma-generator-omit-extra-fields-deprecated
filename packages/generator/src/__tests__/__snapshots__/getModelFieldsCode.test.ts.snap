// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model fields generation: Post 1`] = `"const postKeys = [ 'id', 'createdAt', 'updatedAt', 'published', 'title', 'author', 'authorId',]"`;

exports[`model fields generation: User 1`] = `"const userKeys = [ 'id', 'createdAt', 'email', 'name', 'posts',]"`;
